File: sdks/go/dcp/v2/emergency_revocation.go
// Package v2 — DCP v2.0 Emergency Revocation (Go port).

package v2

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"strings"
)

// EmergencyRevocationTokenPair is the output of GenerateEmergencyRevocationToken.
type EmergencyRevocationTokenPair struct {
	RevocationSecret           string // 64 hex chars
	EmergencyRevocationToken   string // "sha256:<hex>"
}

// GenerateEmergencyRevocationToken produces a secret + commitment pair.
// The secret MUST be stored offline; the commitment is embedded in the passport.
func GenerateEmergencyRevocationToken() EmergencyRevocationTokenPair {
	var secret [32]byte
	_, _ = rand.Read(secret[:])
	sum := sha256.Sum256(secret[:])
	return EmergencyRevocationTokenPair{
		RevocationSecret:         hex.EncodeToString(secret[:]),
		EmergencyRevocationToken: "sha256:" + hex.EncodeToString(sum[:]),
	}
}

// VerifyEmergencyRevocationSecret checks sha256(secret) == commitmentToken.
func VerifyEmergencyRevocationSecret(revocationSecret, commitmentToken string) bool {
	if !strings.HasPrefix(commitmentToken, "sha256:") {
		return false
	}
	expected := commitmentToken[len("sha256:"):]
	secretBytes, err := hex.DecodeString(revocationSecret)
	if err != nil || len(secretBytes) != 32 {
		return false
	}
	sum := sha256.Sum256(secretBytes)
	actual := hex.EncodeToString(sum[:])
	return subtle.ConstantTimeCompare([]byte(actual), []byte(expected)) == 1
}

// BuildEmergencyRevocation constructs an unsigned emergency-revocation request.
func BuildEmergencyRevocation(agentID, humanID, revocationSecret string) map[string]interface{} {
	return map[string]interface{}{
		"type":              "emergency_revocation",
		"agent_id":          agentID,
		"human_id":          humanID,
		"revocation_secret": revocationSecret,
		"timestamp":         utcNowISO(),
		"reason":            "key_compromise_emergency",
	}
}

File: sdks/rust/src/v2/emergency_revocation.rs
//! DCP v2.0 Emergency Revocation — Rust port.

use rand::RngCore;
use serde_json::{json, Value};
use sha2::{Digest, Sha256};

use crate::v2::lifecycle::utc_now_iso_pub as utc_now_iso;

#[derive(Debug, Clone)]
pub struct EmergencyRevocationTokenPair {
    pub revocation_secret: String,          // 64 hex chars
    pub emergency_revocation_token: String, // "sha256:<hex>"
}

/// Generate a (secret, commitment) pair. The secret MUST be stored offline.
pub fn generate_emergency_revocation_token() -> EmergencyRevocationTokenPair {
    let mut bytes = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut bytes);
    let mut hasher = Sha256::new();
    hasher.update(bytes);
    let digest = hasher.finalize();
    EmergencyRevocationTokenPair {
        revocation_secret: hex::encode(bytes),
        emergency_revocation_token: format!("sha256:{}", hex::encode(digest)),
    }
}

/// Constant-time verification that `sha256(secret) == commitment`.
pub fn verify_emergency_revocation_secret(revocation_secret: &str, commitment_token: &str) -> bool {
    let Some(expected_hex) = commitment_token.strip_prefix("sha256:") else {
        return false;
    };
    let Ok(secret_bytes) = hex::decode(revocation_secret) else {
        return false;
    };
    if secret_bytes.len() != 32 {
        return false;
    }
    let mut hasher = Sha256::new();
    hasher.update(&secret_bytes);
    let actual = hex::encode(hasher.finalize());
    // Byte-wise comparison in constant time over fixed-length hex digests.
    ct_eq_str(&actual, expected_hex)
}

fn ct_eq_str(a: &str, b: &str) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff: u8 = 0;
    for (x, y) in a.bytes().zip(b.bytes()) {
        diff |= x ^ y;
    }
    diff == 0
}

/// Build an emergency-revocation request payload (unsigned; gateway validates the secret).
pub fn build_emergency_revocation(agent_id: &str, human_id: &str, revocation_secret: &str) -> Value {
    json!({
        "type": "emergency_revocation",
        "agent_id": agent_id,
        "human_id": human_id,
        "revocation_secret": revocation_secret,
        "timestamp": utc_now_iso(),
        "reason": "key_compromise_emergency",
    })
}
